cpp/src/arrow/scalar.h

```diff
@@ -0,0 +1,143 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Object model for scalar (non-Array) values. Not intended for use with large
+// amounts of data
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/variant.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class Array;
+
+/// \brief Base class for scalar values, representing a single value occupying
+/// an array "slot"
+struct ARROW_EXPORT Scalar {
+  /// \brief Whether the value is valid (not null) or not
+  bool is_valid;
+
+  /// \brief The type of the scalar value
+  std::shared_ptr<DataType> type;
+};
```

Comment:
Doesn't this need a virtual destructor if we're going to be using `shared_ptr<Scalar>`s?

Reply:
Good question. I just added some checks to see if `shared_ptr<Buffer>` in a subclass is properly destructed and it is. I'm happy to add a virtual dtor for good measure, though

Reply:
Ah, I forgot: `shared_ptr`'s block manager is instantiated with the constructed type, so `base_ref` is pointing to a `manager<BinaryScalar>` which calls `~BinaryScalar`. The memory won't leak with `shared_ptr<Scalar>`, but it will leak for `unique_ptr<Scalar>`. Probably best to add the virtual dtor

Reply:
done